nilearn/mass_univariate/tests/test_permuted_least_squares.py

```diff
@@ -14,394 +15,481 @@
-    all_mse_intercept = np.array(all_mse_intercept).reshape(
-        (len(perm_ranges), -1)
+def permuted_ols_with_intercept(tested_var, target_var, n_perm, i):
+    output = permuted_ols(
+        tested_var,
+        target_var,
+        model_intercept=True,
+        n_perm=n_perm,
+        two_sided_test=False,
+        random_state=i,
+        output_type="dict",
     )
-    all_mse_intercept2 = np.array(all_mse_intercept2).reshape(
-        (len(perm_ranges), -1)
+    # pval should not be significant
+    assert_array_less(output["logp_max_t"], 1.0)
+    return output["h0_max_t"]
+
+
+def ks_stat_and_mse(df, h0_intercept):
+    """Run Kolmogorov-Smirnov test and compute Mean Squared Error"""
+    kstest_pval = stats.kstest(h0_intercept, stats.t(df).cdf)[1]
+    mse = mean_squared_error(df=df, h0_intercept=h0_intercept)
+    return kstest_pval, mse
+
+
+def mean_squared_error(df, h0_intercept):
+    return np.mean(
+        (
+            stats.t(df).cdf(np.sort(h0_intercept))
+            - np.linspace(0, 1, h0_intercept.size + 1)[1:]
+        )
+        ** 2
     )
 
+
+def check_ktest_p_values_distribution_and_mse(all_kstest_pvals, all_mse):
     # check that a difference between distributions is not rejected by KS test
+    all_kstest_pvals = np.array(all_kstest_pvals).reshape(
+        (len(PERM_RANGES), -1)
+    )
     assert_array_less(0.01, all_kstest_pvals)
-    assert_array_less(0.01, all_kstest_pvals_intercept)
-    assert_array_less(0.01, all_kstest_pvals_intercept2)
 
     # consistency of the algorithm: the more permutations, the less the MSE
+    all_mse = np.array(all_mse).reshape((len(PERM_RANGES), -1))
     assert_array_less(np.diff(all_mse.mean(1)), 0)
-    assert_array_less(np.diff(all_mse_intercept.mean(1)), 0)
-    assert_array_less(np.diff(all_mse_intercept2.mean(1)), 0)
 
 
-def test_permuted_ols_check_h0_noeffect_signswap(random_state=0):
+def test_permuted_ols_check_h0_noeffect_labelswap_centered_var_no_intercept(
+    random_state=RANDOM_STATE,
+):
+    # create dummy design with no effect
     rng = check_random_state(random_state)
+    target_var = rng.randn(N_SAMPLES, 1)
 
-    # design parameters
-    n_samples = 100
-    n_regressors = 1
+    centered_var = np.arange(N_SAMPLES, dtype="f8").reshape((-1, 1))
+    centered_var -= centered_var.mean(0)
+
+    all_kstest_pvals, all_mse = run_permutations(
+        centered_var, target_var, model_intercept=False
+    )
+
+    check_ktest_p_values_distribution_and_mse(all_kstest_pvals, all_mse)
+
+
+def test_permuted_ols_check_h0_noeffect_labelswap_centered_var_and_intercept(
+    random_state=RANDOM_STATE,
+):
+    # create dummy design with no effect
+    rng = check_random_state(random_state)
+    target_var = rng.randn(N_SAMPLES, 1)
+
+    centered_var = np.arange(N_SAMPLES, dtype="f8").reshape((-1, 1))
+    centered_var -= centered_var.mean(0)
+
+    all_kstest_pvals, all_mse = run_permutations(
+        centered_var, target_var, model_intercept=True
```

Comment:
Isn't this test the same as the previous one with `model_intercept=True` instead of False ? If yes, it could be factorized with a pytest parametrization.

Reply:
oh yeah,true: thanks for pointing it out.